blog: Add post about userns for stateful pods (#42778)
kubernetes/website: content/en/blog/_posts/2023-09-15-userns-stateful-pods/index.md

---
layout: blog

> title: "Kuberntes 1.28: Alpha support to run stateful pods too!"

Reviewer: Try:
```suggestion
title: Run Containers Inside A User Namespace With Kubernetes
```

Author: Removed the 1.28 from the title. But that one is too generic IMHO, changed it to: `title: "User namespaces: Now alpha support to run stateful pods too!"`

date: 2023-09-15
slug: userns-alpha
---

> **Authors:** Rodrigo Campos Catelin (Microsoft), Giuseppe Scrivano (Red Hat), Sascha Grunet (Red Hat)

Reviewer: ```suggestion
**Authors:** Rodrigo Campos Catelin (Microsoft), Giuseppe Scrivano (Red Hat), Sascha Grunert (Red Hat)
```
?

Author: Oops, thanks!

> Kubernetes v1.25 introduced support for user namespaces for only stateless

Reviewer: What's a user namespace? (are you assuming that readers already know?)

Author: Added a quick recap of what they are now, thanks

> pods. In Kubernetes 1.28 we lifted that restriction, after some design changes

Reviewer: Does “we” mean Rodrigo Campos Catelin, Giuseppe Scrivano and Sascha Grunet? If not, reword this.

Author: Fixed

were done in 1.27.

The beauty of this feature is that:

>  * it is trivial to adopt (we need just to set a bool in the pod spec)

Reviewer: ```suggestion
 * it is trivial to adopt (you need just to set a bool in the pod spec)
```

Author: Fixed, thanks

 * doesn't need any changes for **most** applications
 * improves security for **all** Kubernetes users by  _drastically_ enhancing

>    the isolation of containers and mitigating CVEs rated HIGH and CRITICAL.

Reviewer: All CVEs with those ratings? Also, really **all** users? Even people who only run Windows Pods?

Author: Removed the "for all kubernetes users" part

> In this post we will explain the changes in 1.28, show a demo of a CVE rated
> HIGH that is not exploitable with user namespaces, what support you need in the
> runtimes and the Linux kernel to use this and what we are planning for future

Reviewer: ```suggestion
In this post we will explain the changes that arrived in the recent Kubernetes v1.28 release.
We'll explain a demo of a vulnerability rated as HIGH that is not exploitable with user
namespaces; we'll also tell you what support you need in the
```

Author: It is hard when your suggestions include things you said I should remove in this very line in another comment (example here: "we" should be removed, however, your suggestion uses it). It will be simpler to act on suggestions if they are how you think that sentence should end up being.

I changed the text to that but when trying to remove the "we", I ended up creating a list of things this post shows, as the text was getting messy.

> runtimes and the Linux kernel to use this and what we are planning for future

Author: Fixed, thanks

releases.

## Changes in 1.28

> As already mentioned, starting from 1.28 we support stateful pods. This means

Reviewer: Does “we” mean Rodrigo Campos Catelin, Giuseppe Scrivano and Sascha Grunet? If not, reword this.

Author: Fixed, thanks

that pods with user namespaces can use any type of volume, they are no longer
limited to only some volume types as before (see the [volume limitations that

> were present in 1.25][volume-limitations]).

Reviewer: I wouldn't link to the v1.25 docs because that link will stop working in about a year. My rough solution: omit the text inside the parens.

Author: Removed, thanks

While doing this we renamed the feature gate to activate this feature, it is no
longer `UserNamespacesStatelessPodsSupport` but from 1.28 onwards you should
use `UserNamespacesSupport`. There were many changes done and the requirements
on the node hosts changed. So we took the opportunity to rename the feature

> flag.

Reviewer: Does “we” mean Rodrigo Campos Catelin, Giuseppe Scrivano and Sascha Grunet? If not, reword this.

Author: Fixed, thanks

[volume-limitations]: https://v1-25.docs.kubernetes.io/docs/concepts/workloads/pods/user-namespaces/#limitations

## Demo

We've created a demo which exploits [CVE 2022-0492][cve-link] and we show how
the exploit occurs without user namespaces and how it is not possible when user

> namespaces are in use.

Author: Fixed, thanks

> This CVE is rated **HIGH** and allows **a container with no special privileges to

Reviewer: ```suggestion
This vulnerability is rated **HIGH** and allows **a container with no special privileges to
```

Author: Fixed, thanks

read/write to any path on the host** and launch processes as root on the host
too.

> {{< youtube id="" title="User Namespaces mitigating CVE-2022-0492 on Kubernetes">}}

Author: I'll update this soon

> {{< youtube id="" title="User Namespaces mitigating CVE-2022-0492 on Kubernetes">}}

Reviewer: Doesn't look right. Shouldn't there be a non-empty `id`?

> Most applications in containers run as root today, or as the same user (UID
> 65534 is a somewhat popular choice). When this feature is enabled, it switches
> them to run as unprivileged users, with no changes needed in the app.

Reviewer: Try
```suggestion
Most applications in containers run as root today, or as a semi-predictable non-root user (user ID
65534 is a somewhat popular choice). When you run a Pod with containers using a userns,
Kubernetes runs those containers as unprivileged users, with no changes needed in your app.
```

Author: Thanks!

This means two containers running as user 65534 will effectively be mapped to
different users on the host, limiting what they can do to each other in case of
an escape, and if they are running as root, the privileges on the host are
reduced to the one of an unprivileged user.

> [cve-link]: https://unit42.paloaltonetworks.com/cve-2022-0492-cgroups/

Author: The link I pointed explain in more detail how the CVE works, it is a more interesting read IMHO if you want to know more about the CVE.

> [cve-link]: https://unit42.paloaltonetworks.com/cve-2022-0492-cgroups/

Reviewer: How about providing both links, explaining that the second one has more detail? (I don't want to imply that the Palo Alto page is official, or that the Kubernetes project endorses Palo Alto).

> [cve-link]: https://unit42.paloaltonetworks.com/cve-2022-0492-cgroups/

Author: Hmm, I couldn't find a way to mention both things without it looking convulted. The current sentence focus is on how this doesn't happen with userns, explaining both links... I don't know, I couldn't make it fit.

Besides the other link is trivial to get.

## Stack requirements

There are requirements on the Linux kernel version as well as the container
runtime to use this feature.

On Linux you need Linux 6.3 or greater. This is because we rely on a kernel
feature named idmap mounts, and we merged support to use idmap mounts with

> tmpfs in Linux 6.3.

Author: It was Giuseppe and me who wrote the patches and the tests. But no need to highlight that there.

If you are using CRI-O with crun, this is [supported in CRI-O
1.28][CRIO-release]. If you are using CRI-O with runc, this is still not
supported.

> containerd support is currently targeted for containerd 2.0, doesn’t matter if

Author: Fixed, thanks!

you use it with crun or runc.

Please note that contianerd 1.7 added _experimental_ support for user
namespaces as implemented in Kubernetes 1.25 and 1.26. The redesign we did in
1.27 is not supported by containerd 1.7, therefore it only works, in terms of
user namespaces support, with Kubernetes 1.25 and 1.26.

> One limitation present in containerd 1.7 is that it needs to chown the rootfs
> of the container at container startup when using user namespaces. This means it

Reviewer: ```suggestion
One limitation present in containerd 1.7 is that it needs to change the ownership of
the root inode for the filesystem where the runtime has unpacked the container image.
This means it
```
_or_
```suggestion
One limitation present in containerd 1.7 is that it needs to change the ownership of
every file an directory inside the container image, during Pod startup.
This means it
```

Author: The first suggestion is technically wrong (the root inode part). Going with the second one :)

has a storage overhead and can significantly impact the container startup

> latency. For containerd 2.0 we are working on an implementation that will

Reviewer: Does “we” mean Rodrigo Campos Catelin, Giuseppe Scrivano and Sascha Grunet? If not, reword this.

Author: Fixed, thanks

eliminate the startup latency added and the storage overhead. Take this into
account if you plan to use containerd 1.7 with user namespaces in production.

None of these containerd limitations apply to [CRI-O 1.28][CRIO-release].

[CRIO-release]: https://github.com/cri-o/cri-o/releases/tag/v1.28.0

## What’s next?

> For Kubernetes 1.29 we are working with SIG-auth to integrate user namespaces

Reviewer: ```suggestion
Looking ahead to Kubernetes 1.29 we are working with SIG Auth to integrate user namespaces
```

Does “we” mean Rodrigo Campos Catelin, Giuseppe Scrivano and Sascha Grunet? If not, reword this.

Author: Fixed, thanks!

to Pod Security Standards (PSS) and the Pod Security Admission. For the time

> being, we plan to relax checks in PSS policies when user namespaces are in use.

Author: Fixed, thanks

This means that the fields `spec[.*].securityContext` `runAsUser`,
`runAsNonRoot`, `allowPrivilegeEscalation` and `capabilities` will not trigger
a violation if user namespaces are in use. The behavior can be controlled by

> utilizing a new `UserNamespacesPodSecurityStandards` API Server feature gate.

Reviewer: Unless SIG Architecture and SIG Release are committed to delaying v1.29 until that feature gate works like you say, don't say it like this. Frame it as an intent or a hope.

Author: Fixed

---
layout: blog
title: "User namespaces: Now alpha support to run stateful pods too!"

> date: 2023-09-15

Reviewer: Please use
```suggestion
date: 2023-09-13
```
and rename the file accordingly.

Author: Perfect, thanks!

   and GIDs on the host, containers have a harder time to attack each other even
   if they escape the container boundaries. For example, if container A is running
   with different UIDs and GIDs on the host than container B, the operations it
   can do on container B's files and process are limited: only read/write what a
   file allows to others, as it will never have permission for the owner or
   group (the UIDs/GIDs on the host are guaranteed to be different for
   different containers).

 * As the UIDs and GIDs are mapped to unprivileged users on the host, if a
   container escapes the container boundaries, even if it is running as root
   inside the container, it has no privileges on the host. This greatly
   protects what host files it can read/write, which process it can send signals
   to, etc.

Furthermore, capabilities granted are only valid inside the user namespace and
not on the host.

Without using a user namespace a container running as root, in the case of a
container breakout, has root privileges on the node. And if some capabilities
were granted to the container, the capabilities are valid on the host too. None

> of this is true when using user namespaces (modulo bugs, of course :)).

Reviewer: optional change
```suggestion
of this is true when using user namespaces (modulo bugs, of course 🙂).
```

Author: Thanks

## Demo

Rodrigo created a demo which exploits [CVE 2022-0492][cve-link] and shows how
the exploit can occur without user namespaces. He also shows how it is not
possible to use this exploit from a Pod where the containers are using this
feature.

This vulnerability is rated **HIGH** and allows **a container with no special
privileges to read/write to any path on the host** and launch processes as root
on the host too.

{{< youtube id="M4a2b4KkXN8" title="Mitigation of CVE-2022-0492 on Kubernetes by enabling User Namespace support">}}

Most applications in containers run as root today, or as a semi-predictable
non-root user (user ID 65534 is a somewhat popular choice). When you run a Pod
with containers using a userns, Kubernetes runs those containers as unprivileged
users, with no changes needed in your app.

This means two containers running as user 65534 will effectively be mapped to
different users on the host, limiting what they can do to each other in case of
an escape, and if they are running as root, the privileges on the host are
reduced to the one of an unprivileged user.

[cve-link]: https://unit42.paloaltonetworks.com/cve-2022-0492-cgroups/

> ## Stack requirements

Reviewer: ```suggestion
## Node system requirements
```

Author: Fixed

There are requirements on the Linux kernel version as well as the container
runtime to use this feature.

On Linux you need Linux 6.3 or greater. This is because the feature relies on a
kernel feature named idmap mounts, and support to use idmap mounts with tmpfs
was merged in Linux 6.3.

If you are using CRI-O with crun, this is [supported in CRI-O
1.28][CRIO-release] and crun 1.8.7 or greater. If you are using CRI-O with runc,
this is still not supported.

containerd support is currently targeted for containerd 2.0; it is likely that
it won't matter if you use it with crun or runc.

> Please note that contianerd 1.7 added _experimental_ support for user

Reviewer: ```suggestion
Please note that containerd 1.7 added _experimental_ support for user
```

Author: Thanks
